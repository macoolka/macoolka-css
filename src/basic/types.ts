import {
    BackgroundSizeProperty,
    BorderBottomWidthProperty,
    BorderLeftWidthProperty,
    BorderRadiusProperty,
    BorderRightWidthProperty,
    BorderTopWidthProperty,
    BorderWidthProperty,
    BottomProperty,
    FontSizeProperty,
    LeftProperty,
    LetterSpacingProperty,
    LineHeightProperty,
    MarginProperty,
    MarginTopProperty,
    MarginBottomProperty,
    MarginLeftProperty,
    MarginRightProperty,
    PaddingProperty,
    PaddingTopProperty,
    PaddingBottomProperty,
    PaddingLeftProperty,
    PaddingRightProperty,
    RightProperty,
    TopProperty,
    FlexBasisProperty,
    OutlineProperty,
    WidthProperty,
    HeightProperty,
    MaxWidthProperty,
    MinWidthProperty,
    MaxHeightProperty,
    MinHeightProperty,
    BaseProps
} from '../base/types';
import {Overwrite} from 'mocoolka-fp/lib/TypeLevel';

export type PxProps = {
    backgroundSize ?: BackgroundSizeProperty<number>,
    borderBottomWidth?: BorderBottomWidthProperty<number>,
    borderLeftWidth?: BorderLeftWidthProperty<number>,
    borderRadius?: BorderRadiusProperty<number>,
    borderRightWidth?: BorderRightWidthProperty<number>,
    borderTopWidth?: BorderTopWidthProperty<number>,
    borderWidth?: BorderWidthProperty<number>,
    bottom?: BottomProperty<number>,
    fontSize?: FontSizeProperty<number>,
    left?: LeftProperty<number>,
    letterSpacing?: LetterSpacingProperty<number|string>,
    lineHeight?: LineHeightProperty<number>,
    margin?: MarginProperty<number>,
    marginTop?: MarginTopProperty<number>,
    marginBottom?: MarginBottomProperty<number>,
    marginLeft?: MarginLeftProperty<number>,
    marginRight?: MarginRightProperty<number>,
    padding?: PaddingProperty<number>,
    paddingTop?: PaddingTopProperty<number>,
    paddingBottom?: PaddingBottomProperty<number>,
    paddingLeft?: PaddingLeftProperty<number>,
    paddingRight?: PaddingRightProperty<number>,
    right?: RightProperty<number>,
    top?: TopProperty<number>,
    flexBasis?: FlexBasisProperty<number>,
    outline?: OutlineProperty<number>,
};
/* 'backgroundSize' | 'borderBottomWidth' | 'borderLeftWidth' | 'borderRadius' | 'borderRightWidth' |
    'borderTopWidth' | 'borderWidth' | 'bottom' | 'fontSize' |
    'left' | 'letterSpacing' | 'lineHeight' | 'margin' | 'marginTop' | 'marginBottom' | 'marginLeft' | 'marginRight' |
    'padding' | 'paddingTop' | 'paddingBottom' | 'paddingLeft' | 'paddingRight' | 'right' | 'top' |
    'flexBasis' | 'outline'; */
export type PxOrPercents = {
    width?: WidthProperty<number>,
    height?: HeightProperty<number>,
    maxWidth?: MaxWidthProperty<number>,
    minWidth?: MinWidthProperty<number>,
    maxHeight?: MaxHeightProperty<number>,
    minHeight?: MinHeightProperty<number>
};
export type Ms = {
    transitionDuration?: BaseProps['transition']|number,
    transitionDelay?: BaseProps['transitionDelay']|number,
    animationDuration?: BaseProps['animationDuration']|number,
    animationDelay?: BaseProps['animationDelay']|number,
};
// 'transitionDuration' | 'transitionDelay' | 'animationDuration' | 'animationDelay';
export type UnitProps = Overwrite<BaseProps, Ms&PxOrPercents&PxProps>;
