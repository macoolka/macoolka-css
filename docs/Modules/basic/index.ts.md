---
title: basic/index.ts
nav_order: 1
parent: Modules
---

# Overview

---

<h2 class="text-delta">Table of contents</h2>

- [Rule (export)](#rule-export)
- [ruleModule (export)](#rulemodule-export)

---

# Rule (export)

**Signature**

```ts
any
```

Added in v0.2.0

# ruleModule (export)

**Signature**

```ts

RuleModule<Rule<UnitNumberProps, {}, BaseProps, {}, {}>, Rule<{}, {}, {}, {}, {}>, {}>

```

Added in v0.2.0
