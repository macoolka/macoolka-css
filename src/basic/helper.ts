import { CssNode, CssProperties} from '../css';
import {  concatMinArray,reverse } from 'mocoolka-fp/lib/Array';
import Record from 'mocoolka-fp/lib/Record';

export const getBreakpoints = (breakpoints: number[]) =>
    breakpoints.map(a => (`@media screen and (max-width: ${a}em)`));

export const media = <T extends CssProperties>(value: CssNode<T>[]) => (breakpoints: number[]): CssNode<T> => {
    const mediaProps = reverse(concatMinArray(getBreakpoints(breakpoints), value));
   // const first: T = head(mediaProps).map(v => v[1]).getOrElse({} as T);
    const last =  mediaProps.map(([name, v]) =>
     ({ [name]: v }));
    return ( { selector: Record<{ [key: string]: T }>()(last) }) as CssNode<T> ;
};
