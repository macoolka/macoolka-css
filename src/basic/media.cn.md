<section class='chapter'>

# 媒体属性

## 媒体属性提供快捷的定义@media。

语法
```js

 mkMedia: [
   { width: '10px' },
   { width: '300px' },
   { width: '700' },
   { width: '800' }
 ]

```
生成

```css
@media screen and (max-width: 120em) {
  width: 700;
}
@media screen and (max-width: 93em) {
  width: 300px;
}
@media screen and (max-width: 80em) {
  width: 100%;
}
@media screen and (max-width: 50em) {
  width: 10px;
}
```

* 数组中的4个值分别对应4个屏幕宽度
* 值可以是属性也可以为选择器

</section>