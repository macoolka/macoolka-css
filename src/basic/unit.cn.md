<section class='chapter'>

# 单位属性

## 单位属性是对基本属性的扩展。

单位属性主要目的是对没有写明单位的属性赋缺省单位。

单位属性覆盖原始属性，其它属性在单位属性上进行扩展。

例如
```css
width:5 
```
等于
```css
width:5px 
```

<article>

## 缺省单位px

<section class="cols2">
<div>

### 包含的属性名

* backgroundSize
* borderBottomWidth
* borderLeftWidth
* borderRadius
* borderRightWidth,
* borderTopWidth
* borderWidth
* bottom
* fontSize
* left
* letterSpacing
* lineHeight
* margin
* marginTop
* marginBottom
* marginLeft
* marginRight
* padding
* paddingTop
* paddingBottom
* paddingLeft
* paddingRight
* right
* top
* flexBasis

</div>
<div>

例如
```css
width:5 
```
等于
```css
width:5px 
```

</div>

</section>

</article>

<article>

## 缺省单位px或%

* 值在0-1的范围内用%
* 大于1用px

<section class="cols2">
<div>

## 包含的属性名

* width
* height
* maxWidth
* minWidth
* maxHeight
* minHeight

</div>
<div>

例如
```css
width:0.5 
```
等于
```css
width:50% 
```

</div>

</section>

</article>

<article>

## 缺省单位ms


<section class="cols2">
<div>

### 包含的属性名

* transitionDuration
* transitionDelay
* animationDuration
* animationDelay

</div>
<div>

例如
```css
transitionDelay: 500 
```
等于
```css
transitionDelay: 500ms 
```

</div>

</section>

</article>
   

</section>