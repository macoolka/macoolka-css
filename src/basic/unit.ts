/**
 * convert number type to px or precent type
 * @rule
 */
import { BaseProps } from '../base/types';
import { ObjectGet } from 'mocoolka-fp/lib/TypeLevel';
import { Rule, Input as RuleInput } from '../base/rule';
import { isNumber } from 'mocoolka-fp/lib/predicate';
import { getDictionaryMonoid, fold } from 'mocoolka-fp/lib/Monoid';
import { getObjectSemigroup } from 'mocoolka-fp/lib/Semigroup';
import R from 'mocoolka-fp/lib/Record';
const M = getDictionaryMonoid(getObjectSemigroup<any>());

type UnitCssPropeties<U extends keyof BaseProps, T> = ObjectGet<BaseProps, U, T>;
type PxProps = 'backgroundSize' | 'borderBottomWidth' | 'borderLeftWidth' | 'borderRadius' | 'borderRightWidth' |
    'borderTopWidth' | 'borderWidth' | 'bottom' | 'fontSize' |
    'left' | 'letterSpacing' | 'lineHeight' | 'margin' | 'marginTop' | 'marginBottom' | 'marginLeft' | 'marginRight' |
    'padding' | 'paddingTop' | 'paddingBottom' | 'paddingLeft' | 'paddingRight' | 'right' | 'top' |
    'flexBasis'|'outline';
type PxOrPercents = 'width' | 'height' | 'maxWidth' | 'minWidth' | 'maxHeight' | 'minHeight';
export type Props = UnitCssPropeties<PxProps | PxOrPercents, number>;
export const foldUnit = R<Props>();
const pxOrPercent = (n: any): any => !isNumber(n) || n > 1 ? px(n) : (n * 100) + '%';
const px = (n: any): any => isNumber(n) ? n + 'px' : n;

export const pxProps: Array<keyof Props> = [
    'backgroundSize', 'borderBottomWidth', 'borderLeftWidth', 'borderRadius', 'borderRightWidth',
    'borderTopWidth', 'borderWidth', 'bottom', 'fontSize',
    'left', 'letterSpacing', 'lineHeight', 'margin', 'marginTop', 'marginBottom', 'marginLeft', 'marginRight',
    'padding', 'paddingTop', 'paddingBottom', 'paddingLeft', 'paddingRight', 'right', 'top',
    'flexBasis',
];
export const pxOrPercentProps = ['width', 'height', 'maxWidth', 'minWidth', 'maxHeight', 'minHeight'];

const pxArray = pxProps.map((a) => ({
    [a]: (v: any) => ({ [a]: px(v) }),
}));
const pxOrPercentArray = pxOrPercentProps.map((a) => ({
    [a]: (v: any) => ({ [a]: pxOrPercent(v) }),
}));
export type UnitProps = RuleInput<Props, BaseProps>;
export const rule = {rule: fold(M)(pxArray.concat(pxOrPercentArray))} as Rule<Props, {}, BaseProps>;
