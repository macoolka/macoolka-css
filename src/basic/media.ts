
/**
 * convert media props type to props
 * @rule
 */
import { CssProperties, Rule, CssNode } from '../css';

import {media} from './helper';
export type Theme= {
    breakpointUnit: number,
breakpoints: {
    sm: number
    md: number
    lg: number
    xl: number
},
};
export const theme: Theme = {
    breakpointUnit: 16,
    breakpoints: {
        sm: 800,
        md: 1280,
        lg: 1480,
        xl: 1920,
    },
};
const getBreakpoints = (t: Theme): number[] =>
[theme.breakpoints.sm, theme.breakpoints.md, theme.breakpoints.lg, theme.breakpoints.xl]
.map(a => (Math.round(a / t.breakpointUnit)));
/**
 * convert number type to px or precent
 * @getter
 */

export interface Props<P extends CssProperties>  {
    mkMedia?: CssNode<P>[];
}

export const rule = <P extends CssProperties, T extends Theme>(): Rule<Props<P>, {}, P, T> => ({
    rule: {
        mkMedia: ({value, theme : t}) => {
           const result = media<P>(value)(getBreakpoints(t));
           return result;
        },
    },
});
