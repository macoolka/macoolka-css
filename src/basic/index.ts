import { CssNode } from '../base/CssNode';
import { Rule, Input as RuleInput, parse as parseRule } from '../base/rule';
import {  UnitProps, rule as unit } from './unit';
import { Props as MediaProps, theme as mediaTheme, rule as mediaRule, Theme as MediaTheme } from './media';
import { BaseProps } from '../base/types';
import { compose } from 'mocoolka-fp/lib/function';
const theme = mediaTheme;
type Theme = MediaTheme;
export {
    theme,
    Theme,
    MediaProps,
    UnitProps,
};
export const parseUnitRule = <I, IEnum extends { [key: string]: string }, T= any,
    >(rules: Rule<I, IEnum, UnitProps, T>) =>
     (t: T) =>
        (i: CssNode<UnitOutput<I&IEnum>>): CssNode<BaseProps> => {
            return compose(parseRule(unit)(t), parseRule(rules)(t))(i);
        };
export type MediaInput<P>= RuleInput<MediaProps<P>, P >;
export type UnitOutput<P>= RuleInput<P, UnitProps>;
export const parseUnit = <P, T= any>(t: T) =>
(i: CssNode<UnitOutput<P>>): CssNode<BaseProps> =>
 parseRule(unit)(t)(i);
export const parseMedia = <P, T= any>(t: T & MediaTheme) =>
 (i: CssNode<MediaInput<P>>): CssNode<P> =>
  parseRule(mediaRule<P>())(t)(i);
