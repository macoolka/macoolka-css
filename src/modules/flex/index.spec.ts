import { parseUnitProp as _parse} from '../../basic';
import {rule} from './index'
const parse = _parse(rule,{});
//import { cssNodeToStringGetter } from '../../base/print';


describe('flex', () => {
/*     it('print flex', () => {
        expect(cssNodeToStringGetter().get(parse({
            mkFlexAlign: 'center', mkFlexAlignV: 'center', mkFlexAlignLines: 'center', mkFlexWrap: 'wrap', mkFlexDirection: 'row'
        }))).toMatchSnapshot();
    }); */
    describe('parse flex', () => {
        it('flexAlign', () => {
            expect(parse({ mkFlexAlign: 'center' })).toEqual({ 'justifyContent': 'center' });
        });
        it('flexAlignV', () => {
            expect(parse({ mkFlexAlignV: 'center' })).toEqual({ 'alignItems': 'center' });
        });
        it('flexAlignLines', () => {
            expect(parse({ mkFlexAlignLines: 'center' })).toEqual({ 'alignContent': 'center' });
        });
        it('flexWrap', () => {
            expect(parse({ mkFlexWrap: 'wrap' })).toEqual({ 'flexWrap': 'wrap' });
        });
        it('flexDirection', () => {
            expect(parse({ mkFlexDirection: 'row' })).toEqual({ 'flexDirection': 'row' });
        });
        expect(parse({
            mkFlexAlign: 'center',
            mkFlexAlignV: 'center',
            mkFlexAlignLines: 'center',
            mkFlexWrap: 'wrap',
            mkFlexDirection: 'row',
            mkFlexItemAlign: 'center',
            mkFlexItemAlignV: 'center',
            mkFlexItemGrow: 1,
            mkFlexItemOrder: 1,
            mkFlexItemShrink: 1,
            mkFlexItemWidth: 1,

        })).toEqual({
            'justifyContent': 'center',
            'alignItems': 'center',
            'alignContent': 'center',
            'flexWrap': 'wrap',
            'flexDirection': 'row',
            'justifySelf': 'center',
            'alignSelf': 'center',
            'flexGrow': 1,
            'order': 1,
            'flexShrink': 1,
            'flexBasis': '1px',
        });
    })
});

