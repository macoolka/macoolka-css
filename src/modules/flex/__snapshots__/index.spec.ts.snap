// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`flex print flex 1`] = `
"justify-content: center;
align-items: center;
align-content: center;
flex-wrap: wrap;
flex-direction: row;"
`;
